cabal-install/src/Distribution/Client/ProjectBuilding/UnpackedPackage.hs

```diff
@@ -126,20 +132,21 @@ import Distribution.Client.ProjectBuilding.PackageFileMonitor
 -- Depending on whether we are installing the package or building it inplace,
 -- the phases will be carried out differently. For example, when installing,
 -- the test, benchmark, and repl phase are ignored.
-data PackageBuildingPhase
-  = PBConfigurePhase {runConfigure :: IO ()}
-  | PBBuildPhase {runBuild :: IO ()}
-  | PBHaddockPhase {runHaddock :: IO ()}
-  | PBInstallPhase
-      { runCopy :: FilePath -> IO ()
-      , runRegister
+data PackageBuildingPhase r where
+  PBConfigurePhase :: {runConfigure :: IO InLibraryLBI} -> PackageBuildingPhase InLibraryLBI
+  PBBuildPhase :: {runBuild :: IO [MonitorFilePath]} -> PackageBuildingPhase ()
+  PBHaddockPhase :: {runHaddock :: IO [MonitorFilePath]} -> PackageBuildingPhase ()
+  PBInstallPhase
+    :: { runCopy :: FilePath -> IO ()
+       , runRegister
           :: PackageDBStack
           -> Cabal.RegisterOptions
           -> IO InstalledPackageInfo
-      }
-  | PBTestPhase {runTest :: IO ()}
-  | PBBenchPhase {runBench :: IO ()}
-  | PBReplPhase {runRepl :: IO ()}
+       }
+    -> PackageBuildingPhase ()
+  PBTestPhase :: {runTest :: IO ()} -> PackageBuildingPhase ()
+  PBBenchPhase :: {runBench :: IO ()} -> PackageBuildingPhase ()
+  PBReplPhase :: {runRepl :: IO ()} -> PackageBuildingPhase ()
```

Comment:
I see, we can adapt `PackageBuildingPhase` to tell us the output of the action using GADTs. Neat.